File: go/objectmodel/origin.go
package objectmodel

// OriginKind is one of the three terms of the origin grammar (SPEC §5.2,
// INV-013).
type OriginKind string

const (
	// OriginYAML — the instance YAML explicitly supplied this value.
	OriginYAML OriginKind = "yaml"
	// OriginSchema — the value materialized from this node's schema default.
	OriginSchema OriginKind = "schema"
	// OriginSealed — the node came from a closed template at a path.
	OriginSealed OriginKind = "sealed"
)

// OriginTerm is a single term of an origin. Template and Path are set only
// for OriginSealed, and INV-015 requires both.
type OriginTerm struct {
	Kind     OriginKind
	Template string
	Path     string
}

// Origin classifies authoring authority (SPEC §5). It is a classification,
// not a history (INV-014): it never accumulates lifecycle events.
type Origin []OriginTerm

// sealedCtx is the (template, path) pair a sealed boundary contributes to
// every origin below it (INV-015).
type sealedCtx struct {
	template string
	path     string
}

func originYAML() Origin   { return Origin{{Kind: OriginYAML}} }
func originSchema() Origin { return Origin{{Kind: OriginSchema}} }

func originSealed(s *sealedCtx) Origin {
	return Origin{{Kind: OriginSealed, Template: s.template, Path: s.path}}
}

// originSealedSchema is truth-table row 4 (SPEC §5.3): the template defined
// and closed the node, and the value came from the template node's own schema
// default. The two terms describe different dimensions, which is why they
// combine where yaml and schema do not (INV-017).
func originSealedSchema(s *sealedCtx) Origin {
	return Origin{{Kind: OriginSealed, Template: s.template, Path: s.path}, {Kind: OriginSchema}}
}

func (o Origin) project() any {
	out := make([]any, 0, len(o))
	for _, t := range o {
		if t.Kind == OriginSealed {
			inner := newOrderedMap()
			inner.set("template", t.Template)
			inner.set("path", t.Path)
			term := newOrderedMap()
			term.set("sealed", inner)
			out = append(out, term)
			continue
		}
		out = append(out, string(t.Kind))
	}
	return out
}

File: rust/src/origin.rs
//! `origin` — authoring authority (SPEC §5).
//!
//! The grammar has exactly four productions and INV-013 says no other form is
//! valid:
//!
//! ```text
//!   [ yaml ]
//! | [ schema ]
//! | [ sealed(template, path) ]
//! | [ sealed(template, path), schema ]
//! ```
//!
//! It is represented here as a closed enum rather than a vector of terms,
//! because a vector invites exactly the mistake the Go implementation made: it
//! validated the terms individually, checked the truth-table exclusions, and
//! never checked the SEQUENCE — so `[yaml, yaml]`, `[schema, schema]` and the
//! fourth production reversed all passed. A type that cannot represent those
//! states does not need a check that rejects them.
//!
//! Rows 5–8 of the §5.3 truth table are the invalid combinations, and they are
//! simply absent from this enum.

/// A sealed term: the closed template and the path within it (INV-015 — both,
/// always; template identity alone is not provenance).
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Sealed {
    pub template: String,
    pub path: String,
}

/// The authoring authority of a node. Four variants, one per production.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Origin {
    /// Row 1 — the instance YAML explicitly supplied this value.
    Yaml,
    /// Row 2 — the value materialized from this node's schema default.
    Schema,
    /// Row 3 — the node and its value came from a closed template.
    Sealed(Sealed),
    /// Row 4 — the template defined and closed the node; the value came from
    /// the template node's own schema default. These combine where `yaml` and
    /// `schema` do not, because `sealed` constrains structural authority while
    /// `schema` describes the value source: different dimensions (INV-017).
    SealedSchema(Sealed),
}

impl Origin {
    /// True when this origin carries a `sealed` term, i.e. authoring is closed
    /// at and below the node (INV-019).
    #[must_use]
    pub fn is_sealed(&self) -> bool {
        matches!(self, Origin::Sealed(_) | Origin::SealedSchema(_))
    }

    /// The sealed term, when there is one.
    #[must_use]
    pub fn sealed(&self) -> Option<&Sealed> {
        match self {
            Origin::Sealed(s) | Origin::SealedSchema(s) => Some(s),
            _ => None,
        }
    }
}
